core/src/main/scala/offheap/Region.scala

```diff
@@ -29,10 +29,18 @@ trait Region extends Allocator {
     checkOpen
     if (Checked.MEMORY) Sanitizer.unregister(id)
   }
-  def reallocate(addr: Addr, size: Size): Addr =
-    throw new UnsupportedOperationException
+  def reallocate(oldAddr: Addr, oldSize: Size, newSize: Size): Addr = {
+    checkOpen
+    if (newSize <= oldSize)
+      oldAddr
```

Comment:
At the moment regions do not store any metadata about allocations made inside them. This makes allocations just a bump of the pointer and nothing more. This grants us nice 3x performance boost over malloc. If we were to support full-fledged free inside we would need much more complicated implementation that tracks allocation metadata and will probably be slower then the one we have now. Apart from that we'll need much more complicated sanitizer. 

Cyclone experimented with idea of supporting free inside region. They call such regions *reaps*. They are briefly described in [Safe Manual Memory Management in Cyclone](http://www.cs.umd.edu/projects/PL/cyclone/scp.pdf). It would be interesting to use their ideas to implement another kind of region allocator if it works out. But that's non-trivial amount of research for post-0.1 timeframe. Would you be interested at having a look at that later? 

I think for now it's best to have simple implementation of reallocate & free that doesn't actually free anything but also behaves sanely without throwing an exception.